crates/ibc/src/clients/ics07_tendermint/client_state/misbehaviour.rs

```diff
@@ -0,0 +1,163 @@
+
+        if header_1.height() < header_2.height() {
+            return Err(Error::InvalidRawMisbehaviour {
+                reason: format!(
+                    "headers1 height is less than header2 height ({} < {})",
+                    header_1.height(),
+                    header_2.height()
+                ),
+            }
+            .into());
+        }
+
+        Ok(())
+    }
+
+    pub fn check_misbehaviour_header(
+        &self,
+        header: &TmHeader,
+        trusted_consensus_state: &TmConsensusState,
+        current_timestamp: Timestamp,
+    ) -> Result<(), ClientError> {
+        // ensure correctness of the trusted next validator set provided by the relayer
+        check_header_trusted_next_validator_set(header, trusted_consensus_state)?;
+
+        // ensure header timestamp is within trusted period from the trusted consensus state
+        {
+            let duration_since_consensus_state = current_timestamp
+                .duration_since(&trusted_consensus_state.timestamp())
+                .ok_or_else(|| ClientError::InvalidConsensusStateTimestamp {
+                    time1: trusted_consensus_state.timestamp(),
+                    time2: current_timestamp,
+                })?;
+
+            if duration_since_consensus_state >= self.trusting_period {
+                return Err(Error::ConsensusStateTimestampGteTrustingPeriod {
+                    duration_since_consensus_state,
+                    trusting_period: self.trusting_period,
+                }
+                .into());
+            }
+        }
+
+        // ensure that 2/3 of trusted validators have signed the new header
+        {
+            let untrusted_state = header.as_untrusted_block_state();
+            let chain_id = self
+                .chain_id
+                .clone()
+                .with_version(header.height().revision_number())
+                .into();
+            let trusted_state =
+                TmHeader::as_trusted_block_state(header, trusted_consensus_state, &chain_id)?;
+            let options = self.as_light_client_options()?;
+
+            self.verifier
+                .verify_commit_against_trusted(&untrusted_state, &trusted_state, &options)
+                .into_result()?;
+        }
+
+        // run the verification checks on the header based on the trusted
+        // consensus state
+        {
+            let untrusted_state = header.as_untrusted_block_state();
+            let chain_id = self.chain_id.clone().into();
+            let trusted_state =
+                header.as_trusted_block_state(trusted_consensus_state, &chain_id)?;
+            let options = self.as_light_client_options()?;
+
+            self.verifier
+                .validate_against_trusted(
+                    &untrusted_state,
+                    &trusted_state,
+                    &options,
+                    current_timestamp.into_tm_time().unwrap(),
+                )
+                .into_result()?;
+        }
```

Comment:
So ultimately it seems like we want to

1. remove this block
2. add `verify_validator_sets` and `verify_commit` checks

?

Reply:
As discussed, for now we will keep the `validate_against_trusted` check here

Reply:
Since we want all the calls in `self.verifier.verify()`, I just call that directly.

[d439574](https://github.com/cosmos/ibc-rs/pull/584/commits/d4395749ddd011f8fea0b23cd721dbf0dfe61232)

Reply:
I opened https://github.com/informalsystems/tendermint-rs/issues/1294 to discuss a new misbehaviour header verification API. 
Maybe open a new issue here and link ^ ?

Reply:
https://github.com/cosmos/ibc-rs/issues/616